lib/src/appengine_internal.dart

```diff
@@ -237,15 +240,49 @@ Future<db.DatastoreDB> _obtainDatastoreService(
 ///
 /// The returned [memcache.Memcache] will be usable within the current service
 /// scope.
-Future<memcache.Memcache> _obtainMemcacheInstance() async {
-  var rawMemcache;
-  try {
-    const int defaultMemcachedPort = 11211;
-    const String testKey = '__memcached_test_key';
-    final testValue = 'fobar-${new DateTime.now()}';
+Future<memcache.Memcache> _obtainMemcacheInstance(Logging logging) async {
+  //
+  // Try to connect to the app engine instance
+  //
+  const hostKey = 'GAE_MEMCACHE_HOST';
+  const portKey = 'GAE_MEMCACHE_PORT';
+
+  var hostVar = Platform.environment[hostKey];
+  var hostPort = Platform.environment[portKey];
+
+  if (hostVar != null && hostPort != null) {
+    var portNumber = int.parse(hostPort);
+    var instance = await _tryMemcacheInstance(hostVar, portNumber, logging);
+    if (instance != null) {
+      return instance;
+    }
+  }
+
+  //
+  // Try to connect to local host
+  //
+  const int defaultMemcachedPort = 11211;
+  var instance =
+      await _tryMemcacheInstance('localhost', defaultMemcachedPort, logging);
+  if (instance != null) {
+    return instance;
+  }
+
+  // We were unable to connect to a memcached server.
+  // Fall back to a dummy NOP memcache instance.
+  final nopMemcache = new nop_memcache_impl.NopMemcacheRpcImpl();
+  return new memcache.Memcache.fromRaw(nopMemcache);
+}
 
-    rawMemcache = new memcache_raw.BinaryMemcacheProtocol(
-        'localhost', defaultMemcachedPort);
+Future<memcache.Memcache> _tryMemcacheInstance(
+    String host, int port, Logging logging) async {
+  var now = new DateTime.now();
+  final String testKey = '__test_key_${now.microsecondsSinceEpoch}';
```

Comment:
Since we may have many instances booting up at the same time – we want to avoid collisions, right?

Reply:
The likelihood of a clash is very minor, but it can't hurt.